File: src/C++SIM/Stat/src/Variance.cc
/*
 * Copyright (C) 1994
 *
 * Department of Computing Science,
 * The University,
 * Newcastle upon Tyne,
 * UK.
 */


#include <errno.h>
#include <iostream.h>
#include <fstream.h>
#include <iomanip.h>

#ifndef ERROR_H_
#  include <Common/Error.h>
#endif

#ifndef VARIANCE_H_
#  include <Stat/Variance.h>
#endif


Variance::Variance () { Variance::reset(); }

Variance::~Variance () {}

void Variance::reset ()
{
    _sqr = 0.0;
    Mean::reset();
}

void Variance::setValue (double value)
{
    _sqr += value*value;
    Mean::setValue(value);
}

double Variance::variance () const
{
    return ((_Number > 1) ? ((_sqr - ((_Sum * _Sum) / _Number)) / (_Number -1)) : 0.0);
}

double Variance::confidence (double)
{
    error_stream << WARNING << "Variance::confidence not implemented yet." << endl;
    return 0.0;
}

Boolean Variance::saveState (ofstream& oFile) const
{
    if (!oFile)
	return FALSE;

    oFile << " " << _sqr;
    return Mean::saveState(oFile);
}

Boolean Variance::saveState (const char* fileName) const
{
    ofstream oFile(fileName, ios::out);
    Boolean result;
    
    if (!oFile)
    {
	error_stream << WARNING << "Variance::saveState - error "
		     << errno << " for file " << fileName << endl;
	return FALSE;
    }
    
    result = saveState(oFile);
    oFile.close();
    return result;
}

Boolean Variance::restoreState (ifstream& iFile)
{
    if (!iFile)
	return FALSE;
    
    iFile >> _sqr;

    return Mean::restoreState(iFile);
}

Boolean Variance::restoreState (const char* fileName)
{
    ifstream iFile(fileName, ios::in);
    Boolean result;
    
    if (!iFile)
    {
	error_stream << WARNING << "Variance::restoreState - error "
		     << errno << " for file " << fileName << endl;
	return FALSE;
    }
    
    result = restoreState(iFile);
    iFile.close();
    return result;
}

ostream& Variance::print (ostream& strm) const
{
    strm << "\nVariance          : " << variance() << endl;
    strm << "Standard Deviation: " << stdDev() << endl;

    return Mean::print(strm);
}

#ifdef NO_INLINES
#  define VARIANCE_CC_
#  include <Stat/Variance.n>
#  undef VARIANCE_CC_
#endif

File: src/JavaSim/classes/arjuna/JavaSim/Statistics/Variance.java
/*
 * Copyright 1990-2008, Mark Little, University of Newcastle upon Tyne
 * and others contributors as indicated 
 * by the @authors tag. All rights reserved. 
 * See the copyright.txt in the distribution for a
 * full listing of individual contributors. 
 * This copyrighted material is made available to anyone wishing to use,
 * modify, copy, or redistribute it subject to the terms and conditions
 * of the GNU Lesser General Public License, v. 2.1.
 * This program is distributed in the hope that it will be useful, but WITHOUT A 
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A 
 * PARTICULAR PURPOSE.  See the GNU Lesser General Public License for more details.
 * You should have received a copy of the GNU Lesser General Public License,
 * v.2.1 along with this distribution; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, 
 * MA  02110-1301, USA.
 * 
 * (C) 1990-2008,
 */

package arjuna.JavaSim.Statistics;

import java.io.*;
import java.lang.Math;

import java.io.FileNotFoundException;
import java.io.IOException;
import java.lang.IllegalArgumentException;

/**
  Used to obtain the variance of the samples given.
  */

public class Variance extends Mean
{
    
public Variance ()
    {
	reset();
    }

    /**
      Add 'value', updating the variance.
      */
    
public void setValue (double value) throws IllegalArgumentException
    {
	_sqr += value*value;
	super.setValue(value);
    }

    /**
      Zero the statistics.
      */
    
public void reset ()
    {
	_sqr = 0.0;
	super.reset();
    }

    /**
      Returns the variance.
      */
    
public double variance ()
    {
	if (_Number > 1)
	    return ((_sqr - ((_Sum * _Sum) / _Number)) / (_Number -1));
	else
	    return 0.0;
    }

    /**
      Returns the standard deviation of the samples.
      */
      
public double stdDev ()
    {
	if (_Number == 0 || variance() <= 0)
	    return 0.0;
	else
	    return Math.sqrt(variance());	
    }

    /**
      Returns the confidence.
      */
      
public double confidence (double value)
    {
	System.out.println("Variance::confidence not implemented yet.");
	return 0.0;	
    }

    /**
      Prints out the statistics information.
      */
    
public void print ()
    {
	System.out.println("Variance          : "+variance());
	System.out.println("Standard Deviation: "+stdDev());

	super.print();
    }

    /**
      Save the state of the histogram to the file named 'fileName'.
      */
    
public boolean saveState (String fileName) throws IOException
    {
	FileOutputStream f = new FileOutputStream(fileName);
	DataOutputStream oFile = new DataOutputStream(f);

	boolean res = saveState(oFile);

	f.close();

	return res;
    }

    /**
      Save the state of the histogram to the stream 'oFile'.
      */
    
public boolean saveState (DataOutputStream oFile) throws IOException
    {
	oFile.writeDouble(_sqr);
	return super.saveState(oFile);	
    }

    /**
      Restore the histogram state from the file 'fileName'.
      */
    
public boolean restoreState (String fileName) throws FileNotFoundException, IOException
    {
	FileInputStream f = new FileInputStream(fileName);
	DataInputStream iFile = new DataInputStream(f);

	boolean res = restoreState(iFile);

	f.close();

	return res;
    }

    /**
      Restore the histogram state from the stream 'iFile'.
      */
    
public boolean restoreState (DataInputStream iFile) throws IOException
    {
	_sqr = iFile.readDouble();

	return super.restoreState(iFile);
    }
    
private double _sqr;

};
